app/mailers/kudo_mailer.rb

```diff
@@ -3,46 +3,55 @@ class KudoMailer < ActionMailer::Base
+  default from: "Archive of Our Own " + "<#{ArchiveConfig.RETURN_ADDRESS}>"
 
-  def kudo_notification(user_id, kudo_id)
-    user = User.find(user_id)
-    kudo = Kudo.find(kudo_id)
-    @pseud = kudo.pseud
-    @commentable = kudo.commentable
-    I18n.with_locale(Locale.find(user.preference.preferred_locale).iso) do
-      mail(
-        to: user.email,
-        subject: "[#{ArchiveConfig.APP_SHORT_NAME}] Kudos on " + @commentable.commentable_name.gsub("&gt;", ">").gsub("&lt;", "<")
-      )
-    end
-    ensure
-      I18n.locale = I18n.default_locale
-  end
-  
   # send a batched-up notification 
   # user_kudos is a hash of arrays converted to JSON string format
-  # [commentable_type]_[commentable_id] => [array of users who left kudos with the last entry being "# guests" if any]
+  # [commentable_type]_[commentable_id] =>
+  #   names: [array of users who left kudos with the last entry being "# guests" if any]
+  #   guest_count: number of guest kudos
   def batch_kudo_notification(user_id, user_kudos)
     @commentables = []
     @kudo_givers = {}
     user = User.find(user_id)
     kudos_hash = JSON.parse(user_kudos)
-    kudos_hash.each_pair do |commentable_info, kudo_givers|
-      commentable_type, commentable_id = commentable_info.split("_")
-      commentable = commentable_type.constantize.find_by_id(commentable_id)
-      next unless commentable
-      @commentables << commentable
-      @kudo_givers[commentable_info] = kudo_givers
-    end
+
     I18n.with_locale(Locale.find(user.preference.preferred_locale).iso) do
+      kudos_hash.each_pair do |commentable_info, kudo_givers_hash|
+        # Parse the key to extract the type and id of the commentable - skip if no commentable
+        commentable_type, commentable_id = commentable_info.split('_')
+        commentable = commentable_type.constantize.find_by_id(commentable_id)
+        next unless commentable
+
+        # If we have a commentable, extract names and process guest kudos text - skip if no kudos givers
+        names = kudo_givers_hash['names']
+        guest_count = kudo_givers_hash['guest_count']
+        kudo_givers = {}
+
+        if !names.nil? && names.size > 0
+          kudo_givers = names
+          kudo_givers << guest_kudos(guest_count) unless guest_count == 0
+        else
+          kudo_givers << guest_kudos(guest_count).capitalize unless guest_count == 0
+        end
+        next if kudo_givers.empty?
+
+        @commentables << commentable
+        @kudo_givers[commentable_info] = kudo_givers.to_sentence
+      end
       mail(
         to: user.email,
-        subject: "[#{ArchiveConfig.APP_SHORT_NAME}] You've got kudos!"
+        subject: "[#{ArchiveConfig.APP_SHORT_NAME}] #{t 'mailer.kudos.youhave'}"
       )
     end
     ensure
       I18n.locale = I18n.default_locale
   end
 
+  def guest_kudos(guest_count)
+    if guest_count.to_i == 1
+      return "#{t 'mailer.kudos.guest'}"
+    end
+    return "#{guest_count} #{t 'mailer.kudos.guests'}"
```

Comment:
I've found that you get used to the absence of `return` keywords when you start seeing the methods as functions that "are" their return values, rather than separate procedures that have to "return" a value.

The way you've formatted this, the first `return` is indeed necessary, but I'd personally find the logic easier to follow if the whole thing was one `if` statement:

```
if guest_count.to_i == 1
  "#{t 'mailer.kudos.guest'}"
else
  "#{guest_count} #{t 'mailer.kudos.guests'}"
end
```